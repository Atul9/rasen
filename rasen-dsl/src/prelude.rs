//! Exports everything you need to have in scope for the DSL to work

pub use rasen::prelude::*;

pub use types::*;
pub use value::*;
pub use shader::*;
pub use operations::*;
