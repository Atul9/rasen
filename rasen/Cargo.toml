[package]
name = "rasen"
version = "0.11.0"
authors = ["l3ops <rasen@leops.me>"]
description = "Build a SPIR-V module from a data flow graph"
documentation = "https://docs.rs/rasen/"
homepage = "https://github.com/leops/rasen#readme"
repository = "https://github.com/leops/rasen"
readme = "README.md"
license = "MIT"
build = "build.rs"
keywords = [
    "SPIRV",
    "Vulkan",
    "GLSL",
]
categories = [
    "development-tools",
    "rendering",
]

[badges]
appveyor = { repository = "leops/rasen" }
travis-ci = { repository = "leops/rasen" }

[features]
default = []

[dependencies]
petgraph = "0.4.5"
error-chain = "0.10.0"
spirv_headers = "1.2.2"
rspirv = "0.5.1"
fnv = "1.0.6"
clippy = { version = "0.0.177", optional = true }

[build-dependencies]
quote = "0.3.15"
clippy = { version = "0.0.177", optional = true }

[dev-dependencies]
pretty_assertions = "0.5.1"
