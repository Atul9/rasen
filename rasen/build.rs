#![recursion_limit = "128"]
#![feature(inclusive_range_syntax)]
#![cfg_attr(feature="clippy", feature(plugin))]
#![cfg_attr(feature="clippy", plugin(clippy))]

#[macro_use]
extern crate quote;

use std::env;
use std::fs::File;
use std::io::Write;
use std::path::Path;

const INTS: [(&'static str, &'static str, &'static str); 3] = [
    ("Bool", "b", "bool"),
    ("Int", "i", "i32"),
    ("UInt", "u", "u32"),
];
const FLOATS: [(&'static str, &'static str, &'static str); 2] = [
    ("Float", "", "f32"),
    ("Double", "d", "f64"),
];

fn types(out_dir: &str) {
    let builder = quote::Ident::from("$builder");
    let constant = quote::Ident::from("$constant");

    let mut const_types = Vec::new();
    let mut from_string_arms = Vec::new();
    let mut typed_variants = Vec::new();
    let mut type_name_arms = Vec::new();
    let mut register_constant_arms = Vec::new();

    for &(name, _, ty) in INTS.iter().chain(FLOATS.iter()) {
        let const_name = quote::Ident::from(name.to_string().to_uppercase());
        let glsl_name = name.to_string().to_lowercase();
        let name = quote::Ident::from(name);

        from_string_arms.push(quote! {
            #glsl_name => TypeName::#const_name,
        });

        type_name_arms.push(quote! {
            TypedValue::#name(..) => TypeName::#const_name
        });

        let register_val = match ty {
            "bool" => quote! {
                if val {
                    #builder.module.types_global_values.push(
                        Instruction::new(
                            Op::ConstantTrue,
                            Some(res_type),
                            Some(res_id),
                            Vec::new()
                        )
                    );
                } else {
                    #builder.module.types_global_values.push(
                        Instruction::new(
                            Op::ConstantFalse,
                            Some(res_type),
                            Some(res_id),
                            Vec::new()
                        )
                    );
                }
            },

            "i32" | "u32" => quote! {
                #builder.module.types_global_values.push(
                    Instruction::new(
                        Op::Constant,
                        Some(res_type),
                        Some(res_id),
                        vec![
                            Operand::LiteralInt32(val as u32),
                        ]
                    )
                );
            },

            "f32" => quote! {
                #builder.module.types_global_values.push(
                    Instruction::new(
                        Op::Constant,
                        Some(res_type),
                        Some(res_id),
                        vec![
                            Operand::LiteralFloat32(val),
                        ]
                    )
                );
            },
            "f64" => quote! {
                #builder.module.types_global_values.push(
                    Instruction::new(
                        Op::Constant,
                        Some(res_type),
                        Some(res_id),
                        vec![
                            Operand::LiteralFloat64(val),
                        ]
                    )
                );
            },

            _ => unreachable!()
        };

        register_constant_arms.push(quote! {
            TypedValue::#name(val) => {
                let res_type = #builder.register_type(#constant.to_type_name());
                let res_id = #builder.get_id();
                #register_val
                Ok(res_id)
            },
        });

        let ty = quote::Ident::from(ty);
        typed_variants.push(quote! {
            #name(#ty)
        });
    }

    for size in 2u32..=4u32 {
        for &(name, prefix, ty) in INTS.iter().chain(FLOATS.iter()) {
            let type_variant = format!("{}Vec{}", prefix.to_string().to_uppercase(), size);
            let const_name = quote::Ident::from(type_variant.to_uppercase());
            let glsl_name = type_variant.to_lowercase();
            let type_variant = quote::Ident::from(type_variant);

            let const_ty = quote::Ident::from(name.to_string().to_uppercase());
            let name = quote::Ident::from(name);

            const_types.push(quote! {
                pub const #const_name: &'static TypeName = &TypeName::Vec(#size, TypeName::#const_ty);
            });

            from_string_arms.push(quote! {
                #glsl_name => TypeName::#const_name,
            });

            let tuple_fields: Vec<_> = (0..size).map(|_| quote::Ident::from(ty)).collect();
            typed_variants.push(quote! {
                #type_variant( #( #tuple_fields ),* )
            });

            type_name_arms.push(quote! {
                TypedValue::#type_variant(..) => TypeName::#const_name
            });

            let fields: Vec<_> = (0..size).map(|i| quote::Ident::from(format!("f{}", i))).collect();
            let field_ids: Vec<_> = (0..size).map(|i| quote::Ident::from(format!("f{}_id", i))).collect();

            let register_fields: Vec<_> =
                field_ids.iter().zip(fields.iter())
                    .map(|(id, field)| quote! {
                        let #id = #builder.register_constant(&TypedValue::#name(#field))?;
                    })
                    .collect();

            register_constant_arms.push(quote! {
                TypedValue::#type_variant(#( #fields ),*) => {
                    #( #register_fields )*

                    let res_type = #builder.register_type(#constant.to_type_name());
                    let res_id = #builder.get_id();

                    #builder.module.types_global_values.push(
                        Instruction::new(
                            Op::ConstantComposite,
                            Some(res_type),
                            Some(res_id),
                            vec![
                                #( Operand::IdRef( #field_ids ) ),*
                            ]
                        )
                    );

                    Ok(res_id)
                },
            });
        }

        for &(_, prefix, ty) in &FLOATS {
            let ty = quote::Ident::from(ty);
            let const_ty = quote::Ident::from(format!("{}VEC{}", prefix.to_string().to_uppercase(), size));

            let type_variant = format!("{}Mat{}", prefix.to_string().to_uppercase(), size);
            let const_ident = quote::Ident::from(type_variant.to_uppercase());
            let glsl_type = type_variant.to_lowercase();
            let type_variant = quote::Ident::from(type_variant);

            const_types.push(quote! {
                pub const #const_ident: &'static TypeName = &TypeName::Mat(#size, TypeName::#const_ty);
            });

            from_string_arms.push(quote! {
                #glsl_type => TypeName::#const_ident,
            });

            let arr_size = (size * size) as usize;
            typed_variants.push(quote! {
                #type_variant([#ty; #arr_size])
            });

            type_name_arms.push(quote! {
                TypedValue::#type_variant(..) => TypeName::#const_ident
            });
        }
    }

    let type_name_impl = quote! {
        impl TypeName {
            #( #const_types )*

            #[inline]
            pub fn from_string(ty: &str) -> Option<&'static TypeName> {
                Some(match ty {
                    #( #from_string_arms )*
                    _ => return None
                })
            }
        }
    };

    let typed_value = quote! {
        /// Holder for a GLSL value and a type
        #[derive(Debug)]
        pub enum TypedValue {
            #( #typed_variants ),*
        }
    };

    let typed_value_impl = quote! {
        impl TypedValue {
            #[inline]
            pub fn to_type_name(&self) -> &'static TypeName {
                match *self {
                    #( #type_name_arms ),*
                }
            }
        }
    };

    let path_types = Path::new(out_dir).join("types.rs");
    let mut f_types = File::create(&path_types).unwrap();

    write!(f_types, "{}\n{}\n{}",
        type_name_impl, typed_value, typed_value_impl
    ).unwrap();

    let register_constant = quote! {
        match *#constant {
            #( #register_constant_arms )*
            _ => Err(ErrorKind::UnsupportedConstant(#constant.to_type_name()))
        }
    };

    let path_builder = Path::new(out_dir).join("builder.rs");
    let mut f_builder = File::create(&path_builder).unwrap();

    write!(f_builder,
        "macro_rules! impl_register_constant {{\n( $builder:expr, $constant:expr ) => {{\n{}\n}};\n}}",
        register_constant
    ).unwrap();
}

const NODES: [(&'static str, &'static str, &'static str); 24] = [(
    "Normalize",
    "Normalize a vector",
    "Takes a single parameter"
), (
    "Add",
    "Add some values",
    "This node takes at least 2 parameters (left-associative)"
), (
    "Subtract",
    "Subtract a value from another",
    "This node takes at least 2 parameters (left-associative)"
), (
    "Multiply",
    "Multiply some values",
    "This node takes at least 2 parameters (left-associative)"
), (
    "Divide",
    "Divide a value by another",
    "This node takes at least 2 parameters (left-associative)"
), (
    "Modulus",
    "Compute the modulus of a value by another",
    "Takes 2 parameters"
), (
    "Clamp",
    "Clamp a value in a range",
    "Takes 3 parameters: the value to be clamped, the minimum, and the maximum"
), (
    "Dot",
    "Compute the dot product of 2 vectors",
    "Takes 2 parameters"
), (
    "Cross",
    "Compute the cross product of 2 vectors",
    "Takes 2 parameters"
), (
    "Floor",
    "Round a number to the largest lower or equal integer",
    "Takes a single parameter"
), (
    "Ceil",
    "Round a number to the nearest integer",
    "Takes a single parameter"
), (
    "Round",
    "Round a number to the smallest higher or equal integer",
    "Takes a single parameter"
), (
    "Sin",
    "Compute the sinus of an angle in radians",
    "Takes a single parameter"
), (
    "Cos",
    "Compute the cosinus of an angle in radians",
    "Takes a single parameter"
), (
    "Tan",
    "Compute the tangent of an angle in radians",
    "Takes a single parameter"
), (
    "Pow",
    "Raise a number to a power",
    "Takes 2 parameters"
), (
    "Min",
    "Returns the smallest value of all its arguments",
    "This node takes at least 2 parameters"
), (
    "Max",
    "Return the greatest value of all its arguments",
    "This node takes at least 2 parameters"
), (
    "Length",
    "Computes the length of a vector",
    "Takes a single parameter"
), (
    "Distance",
    "Computes the distance between 2 points",
    "Takes 2 parameters"
), (
    "Reflect",
    "Reflect a vector against a surface normal",
    "Takes 2 parameters"
), (
    "Refract",
    "Computes the refraction of a vector using a surface normal and a refraction indice",
    "Takes 3 parameters"
), (
    "Mix",
    "Computes a linear interpolation between two values",
    "Takes 3 parameters"
), (
    "Sample",
    "Samples a texture usinga coordinates vector",
    "Takes 2 parameters: the texture sampler and the coordinates"
)];

fn nodes(out_dir: &str) {
    let mut node_variants = Vec::new();
    let mut to_string_arms = Vec::new();
    let mut from_string_arms = Vec::new();

    for &(name, desc, params) in &NODES {
        let ident = quote::Ident::from(name.to_string());
        node_variants.push(quote! {
            #[doc = #desc]
            #[doc = ""]
            #[doc = #params]
            #ident
        });

        to_string_arms.push(quote! {
            Node::#ident => #name
        });

        from_string_arms.push(quote! {
            #name => Node::#ident
        });
    }

    let node = quote! {
        /// All the supported operations
        #[derive(Debug)]
        pub enum Node {
            /// Create an input with a location and a type
            ///
            /// Incoming values from other nodes are ignored
            Input(u32, &'static TypeName),

            /// Create a uniform with a location and a type
            ///
            /// Incoming values from other nodes are ignored
            Uniform(u32, &'static TypeName),

            /// Create an output with a location and a type
            ///
            /// Doesn't need to be an output of the graph, but all the outputs should use this type
            Output(u32, &'static TypeName),

            /// Declare a new constant
            ///
            /// Incoming values from other nodes are ignored
            Constant(TypedValue),

            /// Build a composite object (only vectors are supported at the moment)
            ///
            /// Uses 2, 3 or 4 arguments depending on the specified output type
            Construct(&'static TypeName),

            /// Extract a value from a composite object
            ///
            /// Takes a single argument (only vector types are supported)
            Extract(u32),

            #( #node_variants ),*
        }

        impl Node {
            /// Get the name of this node
            #[inline]
            pub fn to_string(&self) -> &'static str {
                match *self {
                    Node::Input(..) => "Input",
                    Node::Uniform(..) => "Uniform",
                    Node::Output(..) => "Output",
                    Node::Constant(..) => "Constant",
                    Node::Construct(..) => "Construct",
                    Node::Extract(..) => "Extract",
                    #( #to_string_arms ),*
                }
            }

            /// If possible (the node has no payload), construct a node from its name
            #[inline]
            pub fn from_string(str: &str) -> Option<Self> {
                Some(match str {
                    #( #from_string_arms, )*
                    _ => return None,
                })
            }
        }
    };

    let path_types = Path::new(out_dir).join("node.rs");
    let mut f_node = File::create(&path_types).unwrap();

    write!(f_node, "{}", node).unwrap();
}

fn main() {
    let out_dir = env::var("OUT_DIR").unwrap();
    types(&out_dir);
    nodes(&out_dir);
}
