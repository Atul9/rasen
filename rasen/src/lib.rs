//! Build a SPIR-V module from an operation graph
//!
//! This library lets you define a shader module as a graph (using the `petgraph` library) of
//! `Node`, describing the operations needed to obtain the outputs of the shader.
//!
//! ```
//! # extern crate rasen;
//! # use rasen::prelude::*;
//! # fn main() {
//!     let mut graph = Graph::default();
//!
//!     // A vec3 input at location 0
//!     let normal = graph.add_node(Node::Input(0, TypeName::VEC3));
//!
//!     // Some ambient light constants
//!     let min_light = graph.add_node(Node::Constant(TypedValue::Float(0.1)));
//!     let max_light = graph.add_node(Node::Constant(TypedValue::Float(1.0)));
//!     let light_dir = graph.add_node(Node::Constant(TypedValue::Vec3(0.3, -0.5, 0.2)));
//!
//!     // The Material color (also a constant)
//!     let mat_color = graph.add_node(Node::Constant(TypedValue::Vec4(0.25, 0.625, 1.0, 1.0)));
//!
//!     // Some usual function calls
//!     let normalize = graph.add_node(Node::Normalize);
//!     let dot = graph.add_node(Node::Dot);
//!     let clamp = graph.add_node(Node::Clamp);
//!     let multiply = graph.add_node(Node::Multiply);
//!
//!     // And a vec4 output at location 0
//!     let color = graph.add_node(Node::Output(0, TypeName::VEC4));
//!
//!     // Normalize the normal
//!     graph.add_edge(normal, normalize, 0);
//!
//!     // Compute the dot product of the surface normal and the light direction
//!     graph.add_edge(normalize, dot, 0);
//!     graph.add_edge(light_dir, dot, 1);
//!
//!     // Restrict the result into the ambient light range
//!     graph.add_edge(dot, clamp, 0);
//!     graph.add_edge(min_light, clamp, 1);
//!     graph.add_edge(max_light, clamp, 2);
//!
//!     // Multiply the light intensity by the surface color
//!     graph.add_edge(clamp, multiply, 0);
//!     graph.add_edge(mat_color, multiply, 1);
//!
//!     // Write the result to the output
//!     graph.add_edge(multiply, color, 0);
//!
//! #   #[allow(unused_variables)]
//!     let bytecode = build_program(&graph, ShaderType::Fragment).unwrap();
//!     // bytecode is now a Vec<u8> you can pass to Vulkan to create the shader module
//! # }
//! ```
//!
//! On a lower level, you can use the `Builder` struct to build your module by adding instructions
//! directly into it.
//!

#![feature(box_syntax)]
#![cfg_attr(feature="clippy", feature(plugin))]
#![cfg_attr(feature="clippy", plugin(clippy))]
#![cfg_attr(feature="clippy", warn(
    option_unwrap_used, result_unwrap_used,
    single_match_else
))]

extern crate petgraph;
extern crate spirv_headers;
extern crate rspirv;
#[macro_use]
extern crate error_chain;

mod types;
mod operations;
mod builder;
mod node;

pub mod graph;
pub mod errors;
pub mod prelude;
