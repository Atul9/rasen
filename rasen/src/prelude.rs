//! Exports everything you probably want to have in scope to get started with Rasen

use std::convert::TryFrom;

pub use builder::*;
pub use graph::*;
pub use module::Module;
pub use node::*;
pub use types::*;

pub use petgraph::graph::NodeIndex;
pub use spirv_headers::BuiltIn;
pub use spirv_headers::ExecutionModel as ShaderType;

use errors::{Error, Result};

/// Transform a node graph to SPIR-V bytecode
pub fn build_program<'a, I, S>(graph: &'a I, settings: S) -> Result<Vec<u8>>
where
    ModuleBuilder: TryFrom<(&'a I, S), Error = Error>,
{
    let program = ModuleBuilder::try_from((graph, settings))?;
    program.into_binary()
}

/// Transform a node graph to SPIR-V assembly
pub fn build_program_assembly<'a, I, S>(graph: &'a I, settings: S) -> Result<String>
where
    ModuleBuilder: TryFrom<(&'a I, S), Error = Error>,
{
    let program = ModuleBuilder::try_from((graph, settings))?;
    program.into_assembly()
}
