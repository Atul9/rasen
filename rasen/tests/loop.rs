extern crate insta;
extern crate rasen;
extern crate rspirv;

use rasen::prelude::*;

include!("../../tests/update.rs");

#[test]
fn test_build_loop() {
    let mut module = Module::default();

    let func_cond = module.add_function();

    {
        let graph = &mut module[func_cond];
        let input = graph.add_node(Node::Parameter(0, TypeName::FLOAT));
        let cmp = graph.add_node(Node::Greater);
        let ten = graph.add_node(Node::Constant(TypedValue::Float(10.0)));
        let output = graph.add_node(Node::Return);

        graph.add_edge(input, cmp, 0);
        graph.add_edge(ten, cmp, 1);
        graph.add_edge(cmp, output, 0);
    }

    let func_body = module.add_function();

    {
        let graph = &mut module[func_body];
        let input = graph.add_node(Node::Parameter(0, TypeName::FLOAT));
        let add = graph.add_node(Node::Add);
        let one = graph.add_node(Node::Constant(TypedValue::Float(1.0)));
        let output = graph.add_node(Node::Return);

        graph.add_edge(input, add, 0);
        graph.add_edge(one, add, 1);
        graph.add_edge(add, output, 0);
    }

    {
        let graph = &mut module.main;
        let input = graph.add_node(Node::Input(
            0,
            TypeName::FLOAT,
            VariableName::Named(String::from("i_value")),
        ));
        let reduce = graph.add_node(Node::Loop(func_cond, func_body));
        let output = graph.add_node(Node::Output(
            0,
            TypeName::FLOAT,
            VariableName::Named(String::from("o_value")),
        ));

        graph.add_edge(input, reduce, 0);
        graph.add_edge(reduce, output, 0);
    }

    let builder = ModuleBuilder::from_module(
        &module,
        Settings {
            mod_type: ShaderType::Fragment,
            uniforms_name: Some(String::from("uniforms")),
        },
    )
    .expect("from_module");

    let assembly = builder.into_assembly().expect("build");
    assert_spirv_snapshot_matches!("loop.frag", assembly);
}
