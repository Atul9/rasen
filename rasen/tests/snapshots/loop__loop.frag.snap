---
created: "2019-10-25T17:42:55.042347900+00:00"
creator: insta@0.5.2
expression: assembly
source: rasen/tests/loop.rs
---
; SPIR-V
; Version: 1.1
; Generator: Unknown
; Bound: 35
OpCapability Shader
OpMemoryModel Logical GLSL450
OpEntryPoint Fragment %4 "main" %20 %34
OpExecutionMode %4 OriginUpperLeft
OpName %20 "i_value"
OpName %34 "o_value"
OpDecorate %20 Location 0
OpDecorate %34 Location 0
%10 = OpTypeBool
%6 = OpTypeFloat 32
%33 = OpTypePointer Output %6
%34 = OpVariable  %33  Output
%22 = OpTypePointer Function %6
%19 = OpTypePointer Input %6
%20 = OpVariable  %19  Input
%18 = OpTypeFunction %6 %6
%15 = OpConstant  %6  1.0
%12 = OpTypeFunction %10 %6
%8 = OpConstant  %6  10.0
%1 = OpTypeVoid
%2 = OpTypeFunction %1
%4 = OpFunction  %1  None %2
%3 = OpLabel
%21 = OpLoad  %6  %20 None
%23 = OpVariable  %22  Function
OpStore %23 %21 None
OpBranch %24
%24 = OpLabel
OpLoopMerge %25 %26 None
OpBranch %27
%27 = OpLabel
%28 = OpLoad  %6  %23 None
%29 = OpFunctionCall  %10  %5 %28
OpBranchConditional %29 %30 %25
%30 = OpLabel
%31 = OpFunctionCall  %6  %13 %28
OpStore %23 %31 None
OpBranch %26
%26 = OpLabel
OpBranch %24
%25 = OpLabel
%32 = OpLoad  %6  %23 None
OpStore %34 %32 None
OpReturn
OpFunctionEnd
%5 = OpFunction  %10  None %12
%7 = OpFunctionParameter  %6 
%11 = OpLabel
%9 = OpFOrdGreaterThan  %10  %7 %8
OpReturnValue %9
OpFunctionEnd
%13 = OpFunction  %6  None %18
%14 = OpFunctionParameter  %6 
%17 = OpLabel
%16 = OpFAdd  %6  %14 %15
OpReturnValue %16
OpFunctionEnd
